File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer>result = new ArrayList<>();
        for(int i=0;i<nums.length;i++){
            int index = Collections.binarySearch(result, nums[i]);
            if(index < 0){
                index = -index - 1;
            }
            if(index < result.size()){
                result.set(index, nums[i]);
            }else{
                result.add(nums[i]);
            }
            
        }
        return result.size();
    }
}

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int>LIS(nums.size(),1);
        int ans=1;
        for(int i=1;i<nums.size();i++){
            for(int j=0;j<i;j++){
                if(nums[i]>nums[j])
                    LIS[i]=max(LIS[i],LIS[j]+1);
            }
            ans=max(ans,LIS[i]);
        }
        return ans;
    }
};